presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometrySerde.java

```diff
@@ -423,89 +362,164 @@ private static void writePolygon(Geometry geometry, SliceOutput output, boolean
-                        // this is the last part
-                        break;
-                    }
-                    partIndex += polygon.getInteriorRingN(j).getNumPoints();
-                    output.writeInt(partIndex);
-                }
+        int currentPart = 0;
+        int currentPoint = 0;
+        for (int i = 0; i < numGeometries; i++) {
+            Polygon polygon = (Polygon) geometry.getGeometryN(i);
+
+            partIndexes[currentPart] = currentPoint;
+            shellPart[currentPart] = true;
+            int shellPoints = polygon.getExteriorRing().getNumPoints();
+            currentPart++;
+            currentPoint += shellPoints;
+
+            int holesCount = polygon.getNumInteriorRing();
+            for (int holeIndex = 0; holeIndex < holesCount; holeIndex++) {
+                partIndexes[currentPart] = currentPoint;
+                shellPart[currentPart] = false;
+                int holePoints = polygon.getInteriorRingN(holeIndex).getNumPoints();
+                currentPart++;
+                currentPoint += holePoints;
             }
         }
 
-        writeCoordinates(geometry.getCoordinates(), output);
+        for (int partIndex : partIndexes) {
+            output.writeInt(partIndex);
+        }
+
+        Coordinate[] coordinates = geometry.getCoordinates();
+        canonicalizePolygonCoordinates(coordinates, partIndexes, shellPart);
+        writeCoordinates(coordinates, output);
     }
 
-    private static void skipEnvelope(SliceInput input)
+    private static void writeGeometryCollection(DynamicSliceOutput output, Geometry collection)
     {
-        requireNonNull(input, "input is null");
-        input.skip(4 * SIZE_OF_DOUBLE);
+        output.appendByte(GeometryType.GEOMETRY_COLLECTION.code());
+        for (int geometryIndex = 0; geometryIndex < collection.getNumGeometries(); geometryIndex++) {
+            Geometry geometry = collection.getGeometryN(geometryIndex);
+            int startPosition = output.size();
+
+            // leave 4 bytes for the shape length
+            output.appendInt(0);
+            writeGeometry(output, geometry);
+
+            int endPosition = output.size();
+            int length = endPosition - startPosition - Integer.BYTES;
+
+            output.getUnderlyingSlice().setInt(startPosition, length);
+        }
+    }
+
+    private static void writeCoordinate(Coordinate coordinate, SliceOutput output)
+    {
+        output.writeDouble(translateToAVNaN(coordinate.x));
+        output.writeDouble(translateToAVNaN(coordinate.y));
+    }
+
+    private static void writeCoordinates(Coordinate[] coordinates, SliceOutput output)
+    {
+        for (Coordinate coordinate : coordinates) {
+            writeCoordinate(coordinate, output);
+        }
+    }
+
+    private static void writeEnvelope(Envelope envelope, SliceOutput output, boolean isEmpty)
+    {
+        if (isEmpty) {
+            for (int i = 0; i < 4; i++) {
```

Comment:
Please use a constant instead of `4`

Reply:
I know that this is a bad practice to use inline literal value for any numbers but `0` and `1`. But i really have no idea how to call this constant? `RECTANGLE_NUMBER_OF_VERTEXES`? Seems too verbose for such a simple and self explanatory piece of code.